xla/service/gpu/transforms/windowed_einsum_handler.cc

```diff
@@ -299,78 +307,33 @@ absl::Status SetForceDelayForInstruction(HloInstruction* instr,
   return absl::OkStatus();
 }
 
-absl::StatusOr<bool> HandleRsWindowedEinsumLoop(HloComputation* comp,
-                                                int64_t stream_id) {
+absl::StatusOr<bool> HandleRsWindowedEinsumLoop(HloComputation* comp) {
   bool changed = false;
   // If we have a einsum loop with only 1 dot, this means either
   // the loop is not unrolled or only 1 partition is available.
   // It's a no-op in either case.
   if (NumberOfInstructionsInComp(comp, HloOpcode::kDot) <= 1) {
     return changed;
   }
-  for (auto inst : comp->MakeInstructionPostOrder()) {
-    // The dot we'd like to parallelize is consuming the second loop input
-    // as RHS.
-    if (Match(inst, m::Dot())) {
-      // Dispatch the dot to additional compute stream.
-      TF_RETURN_IF_ERROR(UpdateDotAndConsumerConfig(inst, stream_id));
-      ++stream_id;
-      changed = true;
-    }
-
-    // We need to enforce the first collective-permute to be always scheduled
-    // at the beginning of the loop.
-    HloInstruction* matched_cp;
-    if (Match(inst, m::CollectivePermute(
-                        &matched_cp, m::GetTupleElement(m::Parameter(), 2)))) {
-      TF_RETURN_IF_ERROR(
-          SetForceDelayForInstruction(matched_cp, /*force_delay=*/true));
-      changed = true;
-    }
-  }
   // If present, move the dequantization of FP8 operands of the dot into the
   // while loop to allow e.g. gemm_rewriter.cc to fuse the dequantization and
   // dot into an FP8 GEMM.
-  TF_RETURN_IF_ERROR(ShiftDequantizationF8(comp));
+  TF_ASSIGN_OR_RETURN(changed, ShiftDequantizationF8(comp));
   return changed;
 }
 
-absl::StatusOr<bool> HandleAgWindowedEinsumLoop(HloComputation* comp,
-                                                int64_t stream_id) {
+absl::StatusOr<bool> HandleAgWindowedEinsumLoop(HloComputation* comp) {
```

Comment:
Functions `HandleRsWindowedEinsumLoop` and `HandleAgWindowedEinsumLoop` are identical, and pretty much just call `ShiftDequantizationF8`. Remove both, inlining their logic into `WindowedEinsumHandler::Run`.

The `NumberOfInstructionsInComp(comp, HloOpcode::kDot) <= 1` check is no longer done for the call of `UpdateDotAndConsumerConfig`. I don't think this check is useful to guard the call of `ShiftDequantizationF8`. Either remove it outright, or move it to `WindowedEinsumHandler::Run`, only adding the computation to `all_windowed_einsum_loops` if the check holds.

Reply:
updated based on the suggestion